msal/oauth2cli/oauth2.py

```diff
@@ -128,7 +123,11 @@ def _obtain_token(  # The verb "obtain" is influenced by OAUTH2 RFC 6749
 
         _data.update(self.default_body)  # It may contain authen parameters
         _data.update(data or {})  # So the content in data param prevails
-        # We don't have to clean up None values here, because requests lib will.
+        filtered = {k: v for k, v in _data.items() if v is not None}
+        _data.clear()
+        _data.update(filtered)
```

Comment:
I just have it written this was because modifying a dictionary while we are iterating over is not very safe, as mentioned [here](https://stackoverflow.com/questions/6777485/modifying-a-python-dict-while-iterating-over-it/6777632#6777632) and in the official python documentation [here](https://docs.python.org/3/library/stdtypes.html#dictionary-view-objects)

Reply:
Thanks for spending your time on research! I totally enjoy such kind of discussion! That's how we grow.

In this topic, there is a subtle difference. [The pattern you found in that Q&A](https://stackoverflow.com/q/6777485) was not safe, because it changes the data being looped, WHILE still inside the loop.

```python
for k,v in d.iteritems():
    del d[f(k)] # remove some item
```

But that is not applicable to our case here. We did NOT modify the data DURING the loop `{k: v for k, v in _data.items() if v is not None}`. We just have an assignment AFTER the loop, and it is just that the new variable is the same old variable, but when the assignment happens, the loop has already finished.